### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by goderbauer**
This will be used by travis in flutter/plugins and flutter/flutterfire (see https://github.com/flutter/plugins/pull/43 for example)

### CODE REVIEW

**Comment by collinjackson** [comment]
thanks @goderbauer! rslgtm but I'd like to defer to @tvolkert or someone with more commits to flutter_tools on conventions / coding style since this is a big chunk of new code.

@mit-mit might have opinions about the existence of this repo

File: lib/src/format_command.dart [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import 'common.dart';
+
+const String _googleFormatterUrl =
+    'https://github.com/google/google-java-format/releases/download/google-java-format-1.3/google-java-format-1.3-all-deps.jar';
+
+class FormatCommand extends Command {
+  FormatCommand(this.packagesDir) {
+    argParser.addFlag('travis', hide: true);
+    argParser.addOption('clang-format', defaultsTo: 'clang-format');
```
Comment (tvolkert) [inline lib/src/format_command.dart]: Add `help` argument

File: lib/src/format_command.dart [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    if (argResults['travis']) {
+      bool success = _printSummary();
+      if (!success) {
+        throw new ToolExit(1);
+      }
+    }
+  }
+
+  bool _printSummary() {
```
Comment (tvolkert) [inline lib/src/format_command.dart]: Consider renaming to something like `_didModifyAnything` (it does more than printing)

Comment (tvolkert) [inline lib/src/format_command.dart]: `Future<bool>` so you don't have to use `Process.runSync()`

File: lib/src/format_command.dart [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  void _formatObjectiveC() {
+    print('Formatting all .m and .h files...');
+    Iterable<String> hFiles = _getFilesWithExtension(packagesDir, '.h');
+    Iterable<String> mFiles = _getFilesWithExtension(packagesDir, '.m');
+    Process.runSync(argResults['clang-format'],
+        ['-i', '--style=Google']..addAll(hFiles)..addAll(mFiles),
+        workingDirectory: packagesDir.path);
+  }
+
+  void _formatJava(String googleFormaterPath) {
```
Comment (tvolkert) [inline lib/src/format_command.dart]: `Future<Null>` so you can use async process methods.

File: lib/src/format_command.dart [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    print('Formatting all .java files...');
+    Iterable<String> javaFiles = _getFilesWithExtension(packagesDir, '.java');
+    Process.runSync(
+        'java', ['-jar', googleFormaterPath, '--replace']..addAll(javaFiles),
+        workingDirectory: packagesDir.path);
+  }
+
+  void _formatDart() {
```
Comment (tvolkert) [inline lib/src/format_command.dart]: `Future<Null>`

File: lib/src/format_command.dart [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    print('Formatting all .dart files...');
+    Process.runSync('flutter', ['format'], workingDirectory: packagesDir.path);
+  }
+
+  Iterable<String> _getFilesWithExtension(Directory dir, String extension) =>
```
Comment (tvolkert) [inline lib/src/format_command.dart]: `Stream<String>`

File: lib/src/analyze_command.dart [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+class AnalyzeCommand extends Command {
+  AnalyzeCommand(this.packagesDir);
+
+  final Directory packagesDir;
+
+  final name = 'analyze';
+  final description = 'Analyzes all packages.';
+
+  Future run() async {
```
Comment (tvolkert) [inline lib/src/analyze_command.dart]: `Future<Null>`

File: lib/src/main.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    }
+  }
+
+  new CommandRunner('pub global run flutter_plugin_tools',
+      'Productivity utils for hosting multiple plugins within one repository.')
+    ..addCommand(new TestCommand(packagesDir))
+    ..addCommand(new AnalyzeCommand(packagesDir))
+    ..addCommand(new FormatCommand(packagesDir))
+    ..addCommand(new BuildCommand(packagesDir))
+    ..run(args).catchError((ToolExit e) {
```
Comment (tvolkert) [inline lib/src/main.dart]: The use of chaining with non-chaining is a little hard to read here - maybe make this two statements?

File: lib/src/test_command.dart [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+class TestCommand extends Command {
+  TestCommand(this.packagesDir);
+
+  final Directory packagesDir;
+
+  final name = 'test';
+  final description = 'Runs the tests for all packages.';
+
+  Future run() async {
```
Comment (tvolkert) [inline lib/src/test_command.dart]: `Future<Null>`

File: lib/src/test_command.dart [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import 'common.dart';
+
+class TestCommand extends Command {
+  TestCommand(this.packagesDir);
+
+  final Directory packagesDir;
+
+  final name = 'test';
+  final description = 'Runs the tests for all packages.';
```
Comment (tvolkert) [inline lib/src/test_command.dart]: the Dart tests... (we don't really have a story for JUnit or whatever the equivalent is for ObjC)

File: lib/src/test_command.dart [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+  Future run() async {
+    List<String> failingPackages = <String>[];
+    for (Directory packageDir in _listAllPackages(packagesDir)) {
+      String packageName = p.relative(packageDir.path, from: packagesDir.path);
+      if (!new Directory(p.join(packageDir.path, 'test')).existsSync()) {
+        print('\nSKIPPING $packageName - no test subdirectory');
```
Comment (tvolkert) [inline lib/src/test_command.dart]: Preceding newlines again?

File: lib/src/test_command.dart [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+      failingPackages.forEach((String package) {
+        print(' * $package');
+      });
+      throw new ToolExit(1);
+    }
+
+    print('All tests are passing!');
+  }
+
+  Iterable<Directory> _listAllPackages(Directory root) => root
```
Comment (tvolkert) [inline lib/src/test_command.dart]: Stream<Directory>

File: lib/src/build_command.dart [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+class BuildCommand extends Command {
+  BuildCommand(this.packagesDir) {
+    argParser.addFlag('ipa', defaultsTo: Platform.isMacOS);
+    argParser.addFlag('apk');
+  }
+
+  final Directory packagesDir;
+
+  final name = 'build';
+  final description = 'Builds all example apps.';
```
Comment (goderbauer) [inline lib/src/build_command.dart]: This runs as a smoke test an Travis to make sure that the example apps (and with that the plugins (and with that especially the native code of the plugins)) are in a buildable state (most plugins don't have any other tests for their native impl currently).

File: lib/src/build_command.dart [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+  Future run() async {
+    List<String> failingPackages = <String>[];
+    for (Directory example in _getExamplePackages(packagesDir)) {
+      String packageName = p.relative(example.path, from: packagesDir.path);
+
+      if (argResults['ipa']) {
+        print('\nBUILDING IPA for $packageName');
```
Comment (goderbauer) [inline lib/src/build_command.dart]: Each package outputs a lot of logs when build. The newline is to visually separate the individual builds in the log.

File: lib/src/format_command.dart (hunk 5 above)
Comment (goderbauer) [inline lib/src/format_command.dart]: This is only ever processed as an Iterable, so with this change, I'd have to add `toList` calls everywhere. I think that just makes it harder to read.

File: lib/src/test_command.dart (hunk 10 above)
Comment (goderbauer) [inline lib/src/test_command.dart]: Test package prints a lot of logging. Visual separation makes it easier to read.

File: lib/src/format_command.dart [hunk 14]
```diff
@@ -15,35 +15,41 @@ import 'common.dart';
...
     'https://github.com/google/google-java-format/releases/download/google-java-format-1.3/google-java-format-1.3-all-deps.jar';
 
-class FormatCommand extends Command {
+class FormatCommand extends Command<Null> {
   FormatCommand(this.packagesDir) {
     argParser.addFlag('travis', hide: true);
-    argParser.addOption('clang-format', defaultsTo: 'clang-format');
+    argParser.addOption('clang-format',
+        defaultsTo: 'clang-format',
+        help: 'Path to executable of clan-format 3.');
```
Comment (tvolkert) [inline lib/src/format_command.dart]: s/clan-format 3/clang-format/

File: lib/src/format_command.dart [hunk 15]
```diff
@@ -15,35 +15,41 @@ import 'common.dart';
...
+  Future<Null> run() async {
+    final String googleFormatterPath = await _getGogleFormatterPath();
+
+    await _formatDart();
+    await _formatJava(googleFormatterPath);
+    await _formatObjectiveC();
 
     if (argResults['travis']) {
-      bool success = _printSummary();
+      final bool success = await _didModifyAnything();
```
Comment (tvolkert) [inline lib/src/format_command.dart]: Invert the return values, then use `!await _didModifyAnything()`

**Comment by tvolkert** [comment]
LGTM!

**Pull request merged**